### PR TITLE
Add coercion for MPoly constructor.

### LINKED ISSUE #1046: Coerce coefficients for mpoly constructors
If R is a multivariate polynomial ring, then at the moment one has to do `R(QQ.([3, 1]), [[2, 0, 0], [0, 1, 1]])` instead of just `R([3, 1], [[2, 0, 0], [0, 1, 1]])`. If the elements are not in the coefficient ring, they should just be coerced.

Probably needs to be done in Nemo, or we can do it here generically, I don't know yet.

### PR DESCRIPTION
**Pull request by wbhart**
Coerces coefficients in generic version of MPoly constructor.

Fixes #1046

I couldn't think of any distinct rings with the same type that nevertheless have a coercion between them. So I didn't change the code for the Generic MPolyRing type which still raises an exception if the parent of the coefficients passed is not the same as the base ring of the polynomial ring being used to construct the polynomial.

Let me know if you can think of such a case.

### CODE REVIEW

**Comment by thofma** [comment]
Not in AA (yet), but in Nemo:
```
julia> using Nemo

julia> R = ResidueRing(ZZ, 4)
Integers modulo 4

julia> S = ResidueRing(ZZ, 2)
Integers modulo 2

julia> S(R(2))
0
```

**Comment by wbhart** [comment]
Yeah I guess we should fix that one in Nemo.

**Comment by thofma** [comment]
What do you mean?

**Comment by wbhart** [comment]
Oh damn, Julia-1.0 still doesn't allow methods to be added to an abstract type. So this still won't work until the new LTS is available and we can drop 1.0.

**Comment by wbhart** [comment]
See test failures for 1.0.

**Comment by thofma** [comment]
Can you elaborate what 1.0 has to do with the coercion in Nemo?

**Comment by wbhart** [comment]
Sorry, two different conversations, My comment "Oh damn..." was in relation to the test failures in 1.0. Sorry I wasn't clearer.

**Comment by thofma** [comment]
This is breaking Nemo multivariate polynomials:
```julia
MethodError: (::FqNmodMPolyRing)(::Vector{fq_nmod}, ::Vector{Vector{Int64}}) is ambiguous.
```
See the failing `Oscar#master` tests.

**Comment by wbhart** [comment]
Perhaps the best way for me to fix it is to restrict to generic polynomial rings only. If I don't, this would have to be fixed in Nemo.

File: src/MPoly.jl [hunk 1]
```diff
@@ -1100,6 +1100,17 @@ function _map(g, p::MPolyElem, Rx)
    return finish(M)
 end
 
+###############################################################################
+#
+#   Parent object overload (with coercion)
+#
+###############################################################################
+
+function (S::MPolyRing)(b::Vector{U}, m::Vector{Vector{Int}}) where {T <: RingElement, U <: RingElement}
```
Comment (thofma) [inline src/MPoly.jl:1109]:
```suggestion
function (S::MPolyRing)(b::Vector{U}, m::Vector{Vector{T}}) where {T <: RingElement, U <: RingElement, T<:Union{Int64, UInt64, fmpz}}
```
Can you try this? The Nemo versions have this second argument, that is why they are less specific in the second argument.

**Comment by wbhart** [comment]
Clearly the coercion example we discussed above would have to be handled separately in Nemo.

It's all a bit unsatisfying, but this issue with ambiguities just keeps on happening if we define generic functions in AbstractAlgebra to handle the Nemo case.

File: src/MPoly.jl (hunk 1 above)
Comment (thofma) [reply src/MPoly.jl:1109]: And remove the first `T` in the `where` clause.

**Comment by wbhart** [comment]
Hmm, that actually results in a sefault in Julia. I guess we shouldn't do that either:

```
signal (11): Segmentation fault
in expression starting at /home/wbhart/.julia/dev/AbstractAlgebra/test/generic/RationalFunctionField-test.jl:97
_ZN4llvm11Instruction15eraseFromParentEv at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
finalize at /buildworker/worker/package_linux64/build/src/llvm-alloc-opt.cpp:380 [inlined]
runOnFunction at /buildworker/worker/package_linux64/build/src/llvm-alloc-opt.cpp:1518
_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm13FPPassManager11runOnModuleERNS_6ModuleE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
operator() at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:613
_ZN4llvm3orc14IRCompileLayer4emitENS0_29MaterializationResponsibilityENS0_16ThreadSafeModuleE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm3orc31BasicIRLayerMaterializationUnit11materializeENS0_29MaterializationResponsibilityE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession26materializeOnCurrentThreadESt10unique_ptrINS0_19MaterializationUnitESt14default_deleteIS3_EENS0_29MaterializationResponsibilityE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZNSt17_Function_handlerIFvSt10unique_ptrIN4llvm3orc19MaterializationUnitESt14default_deleteIS3_EENS2_29MaterializationResponsibilityEEPS8_E9_M_invokeERKSt9_Any_dataOS6_OS7_ at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession17runOutstandingMUsEv at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession6lookupENS0_10LookupKindERKSt6vectorISt4pairIPNS0_8JITDylibENS0_19JITDylibLookupFlagsEESaIS8_EENS0_15SymbolLookupSetENS0_11SymbolStateENS_15unique_functionIFvNS_8ExpectedINS_8DenseMapINS0_15SymbolStringPtrENS_18JITEvaluatedSymbolENS_12DenseMapInfoISI_EENS_6detail12DenseMapPairISI_SJ_EEEEEEEEESt8functionIFvRKNSH_IS6_NS_8DenseSetISI_SL_EENSK_IS6_EENSN_IS6_SV_EEEEEE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession6lookupERKSt6vectorISt4pairIPNS0_8JITDylibENS0_19JITDylibLookupFlagsEESaIS7_EERKNS0_15SymbolLookupSetENS0_10LookupKindENS0_11SymbolStateESt8functionIFvRKNS_8DenseMapIS5_NS_8DenseSetINS0_15SymbolStringPtrENS_12DenseMapInfoISK_EEEENSL_IS5_EENS_6detail12DenseMapPairIS5_SN_EEEEEE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession6lookupERKSt6vectorISt4pairIPNS0_8JITDylibENS0_19JITDylibLookupFlagsEESaIS7_EENS0_15SymbolStringPtrENS0_11SymbolStateE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession6lookupENS_8ArrayRefIPNS0_8JITDylibEEENS0_15SymbolStringPtrENS0_11SymbolStateE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession6lookupENS_8ArrayRefIPNS0_8JITDylibEEENS_9StringRefENS0_11SymbolStateE at /home/wbhart/julia-1.6.2/bin/../lib/julia/libLLVM-11jl.so (unknown line)
addModule at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:788
jl_add_to_ee at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:1067
jl_add_to_ee at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:1111
jl_add_to_ee at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:1133 [inlined]
_jl_compile_codeinst at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:154
jl_generate_fptr at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:352
jl_compile_method_internal at /buildworker/worker/package_linux64/build/src/gf.c:1970
jl_compile_method_internal at /buildworker/worker/package_linux64/build/src/gf.c:2236 [inlined]
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2229 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:115
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:204
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:155 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:562
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:490
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:490
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:490
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:670
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:877
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:825
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:929
eval at ./boot.jl:360 [inlined]
include_string at ./loading.jl:1116
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
_include at ./loading.jl:1170
include at ./client.jl:444
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:115
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:204
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:155 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:562
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:670
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:877
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:825
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:929
eval at ./boot.jl:360 [inlined]
include_string at ./loading.jl:1116
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
_include at ./loading.jl:1170
include at ./client.jl:444
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:115
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:204
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:155 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:562
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:670
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:877
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:825
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:929
eval at ./boot.jl:360 [inlined]
include_string at ./loading.jl:1116
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
_include at ./loading.jl:1170
include at ./client.jl:444
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:115
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:204
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:155 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:562
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:670
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:877
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:825
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:929
eval at ./boot.jl:360 [inlined]
include_string at ./loading.jl:1116
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
_include at ./loading.jl:1170
include at ./client.jl:444
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:115
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:204
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:155 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:562
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:670
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:877
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:825
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:929
eval at ./boot.jl:360 [inlined]
exec_options at ./client.jl:261
_start at ./client.jl:485
jfptr__start_34281.clone_1 at /home/wbhart/julia-1.6.2/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2419
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
true_main at /buildworker/worker/package_linux64/build/src/jlapi.c:560
repl_entrypoint at /buildworker/worker/package_linux64/build/src/jlapi.c:702
main at /buildworker/worker/package_linux64/build/cli/loader_exe.c:51
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at /home/wbhart/julia-1.6.2/bin/julia (unknown line)
Allocations: 1164259189 (Pool: 1164012277; Big: 246912); GC: 1224
ERROR: Package AbstractAlgebra errored during testing (received signal: 11)
```

File: src/MPoly.jl (hunk 1 above)
Comment (wbhart) [reply src/MPoly.jl:1109]: Actually T is not currently used in the signature. Maybe that is part of the issue too.

Comment (wbhart) [reply src/MPoly.jl:1109]: And I can't put fmpz in the type in AbstractAlgebra.

Comment (thofma) [reply src/MPoly.jl:1109]: Works for me if I do `m::Vector{<: Vector}` (no `T`).

Comment (wbhart) [reply src/MPoly.jl:1109]: Yeah I don't see a way to solve this one. In AbstractAlgebra we only allow UInt (whether 32 or 64 bit) for exponents. This is going to conflict with Nemo whatever we do here.

I tried leaving out the fmpz in the signature in AbstractAlgebra and of course it still gives an ambiguity warning.

Obviously if Nemo wants to allow fmpz and Int exponents, it has to implement those itself. But it cannot also then implement functions with UInt64 exponents that AbstractAlgebra already implements.

Comment (thofma) [reply src/MPoly.jl:1109]: We can just do `m::Vector{<: Vector}`. This works for me.

Comment (thofma) [reply src/MPoly.jl:1109]: Then whatever exponent types the polynomial ring accepts it will still work.

Comment (wbhart) [reply src/MPoly.jl:1109]: Yeah ok, that should be alright. I'll make that change and push after testing locally.

Comment (wbhart) [reply src/MPoly.jl:1109]: I still get the following locally:

```
nmod_mpoly.exponents: Error During Test at /home/wbhart/.julia/packages/Nemo/j2Z5Q/test/flint/nmod_mpoly-test.jl:796
  Test threw exception
  Expression: S([BigInt(0), BigInt(1), BigInt(2)], m) == 1x + 2y
  MethodError: (::NmodMPolyRing)(::Vector{BigInt}, ::Vector{Vector{BigInt}}) is ambiguous. Candidates:
    (S::MPolyRing{T})(b::Vector{U}, m::Vector{var"#s453"} where var"#s453"<:(Vector{T} where T)) where {T<:RingElement, U<:RingElement} in AbstractAlgebra at /home/wbhart/.julia/dev/AbstractAlgebra/src/MPoly.jl:1109
    (R::NmodMPolyRing)(a::Vector{T} where T, b::Array{Vector{T}, 1}) where T in Nemo at /home/wbhart/.julia/packages/Nemo/j2Z5Q/src/flint/nmod_mpoly.jl:1079
  Possible fix, define
    (::NmodMPolyRing)(::Vector{U}, ::Array{Vector{T}, 1}) where {T, U<:RingElement}
  Stacktrace:
   [1] macro expansion
     @ ~/.julia/packages/Nemo/j2Z5Q/test/flint/nmod_mpoly-test.jl:796 [inlined]
   [2] macro expansion
     @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [3] top-level scope
     @ ~/.julia/packages/Nemo/j2Z5Q/test/flint/nmod_mpoly-test.jl:725
```

Comment (wbhart) [reply src/MPoly.jl:1109]: What is the exact signature you are using? I can't get it to work.

Comment (wbhart) [reply src/MPoly.jl:1109]: .... even with the latest Nemo.

**Comment by thofma** [comment]
Superseded by #1073

**Pull request closed without merging**